### PR TITLE
[HIVE-15439] adding support for insert overwite

### PR DESCRIPTION
**Pull request by b-slim**
Add support for SQL statement INSERT OVERWRITE TABLE druid_internal_table.
In order to add this support will need to add new post insert hook to update the druid metadata. Creation of the segment will be the same as CTAS.
https://issues.apache.org/jira/browse/HIVE-15439

### CODE REVIEW

File: metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 1]
```diff
@@ -2180,6 +2180,16 @@ public void addDynamicPartitions(long txnId, String dbName, String tableName,
     client.add_dynamic_partitions(adp);
   }
 
+  @Override
+  public int insertCommit(Table table, boolean overwrite) {
+    try {
+      getHook(table).commitInsert(table, overwrite);
```
Comment (jcamachor) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: If storage handler for the table does not implement the hook, we will get an NPE here. We need a _hook != null_ check, as for the _createTable_ method.

File: metastore/src/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java [hunk 2]
```diff
@@ -1506,6 +1506,8 @@ void addDynamicPartitions(long txnId, String dbName, String tableName, List<Stri
                             DataOperationType operationType)
     throws TException;
 
+  int insertCommit(Table table, boolean overwrite);
```
Comment (jcamachor) [inline metastore/src/java/org/apache/hadoop/hive/metastore/IMetaStoreClient.java]: Add javadoc for the method.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionTimeGranularityOptimizer.java [hunk 3]
```diff
@@ -128,8 +129,16 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
...
-      String segmentGranularity = parseCtx.getCreateTable().getTblProps()
-              .get(Constants.DRUID_SEGMENT_GRANULARITY);
+      String segmentGranularity = null;
+      final Table table = fsOp.getConf().getTable();
+      if (table != null) {
+        // case the statement is an INSERT
+        segmentGranularity = table.getParameters().get(Constants.DRUID_SEGMENT_GRANULARITY);
+      } else {
+        // case the statement is a CREATE TABLE AS
+       segmentGranularity = parseCtx.getCreateTable().getTblProps()
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionTimeGranularityOptimizer.java]: I guess you cannot access the table from fsOp for CTAS? Is that right?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 4]
```diff
@@ -7164,6 +7166,14 @@ protected Operator genFileSinkPlan(String dest, QB qb, Operator input)
     return output;
   }
 
+  private void commitInsert(Table table, boolean overwrite) {
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: Rather change name of method, e.g., to _insert_ or _createInsertDesc_; in the future, the descriptor might be useful beyond commit.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 5]
```diff
@@ -7164,6 +7166,14 @@ protected Operator genFileSinkPlan(String dest, QB qb, Operator input)
...
+    Task<? extends Serializable>[] tasks = new Task[this.rootTasks.size()];
+    tasks = this.rootTasks.toArray(tasks);
+    InsertTableDesc insertCommitHook = new InsertTableDesc(table.getTTable(), overwrite);
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: Change name of object, e.g., to _insertDesc_; in the future, the descriptor might be useful beyond commit.

File: ql/src/java/org/apache/hadoop/hive/ql/plan/DDLWork.java [hunk 6]
```diff
@@ -1185,4 +1193,13 @@ public ShowConfDesc getShowConfDesc() {
   public void setShowConfDesc(ShowConfDesc showConfDesc) {
     this.showConfDesc = showConfDesc;
   }
+
+  @Explain(displayName = "Commit insert operator", explainLevels = { Level.USER, Level.DEFAULT, Level.EXTENDED })
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/plan/DDLWork.java]: Idem as before, make it more general, e.g., "Insert operator".

File: ql/src/java/org/apache/hadoop/hive/ql/plan/InsertTableDesc.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.plan;
+
+import org.apache.hadoop.hive.metastore.api.Table;
+
+@Explain(displayName = "Commit insert operator", explainLevels = { Explain.Level.USER, Explain.Level.DEFAULT, Explain.Level.EXTENDED })
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/plan/InsertTableDesc.java]: display name= "Insert" ? To make it consistent with CreateTableDesc.

File: metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaHook.java [hunk 8]
```diff
@@ -89,4 +89,13 @@ public void rollbackDropTable(Table table)
...
     throws MetaException;
+
+  /**
+   * Called after successfully after INSERT [OVERWRITE] statement is executed.
+   * @param table table definition
+   * @param overwrite true if it is INSERT OVERWRITE
+   *
+   * @throws MetaException
+   */
+  public void commitInsert(Table table, boolean overwrite) throws MetaException;
```
Comment (b-slim) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaHook.java]: @jcamachor do you mean new interface that extends the HiveMetaHook ?

File: ql/src/java/org/apache/hadoop/hive/ql/plan/DDLWork.java (hunk 6 above)
Comment (b-slim) [inline ql/src/java/org/apache/hadoop/hive/ql/plan/DDLWork.java]: ok

File: ql/src/java/org/apache/hadoop/hive/ql/plan/InsertTableDesc.java (hunk 7 above)
Comment (b-slim) [inline ql/src/java/org/apache/hadoop/hive/ql/plan/InsertTableDesc.java]: ok

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionTimeGranularityOptimizer.java (hunk 3 above)
Comment (b-slim) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionTimeGranularityOptimizer.java]: yes

File: metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaHook.java (hunk 8 above)
Comment (b-slim) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaHook.java]: Also should be add the pre/rollback even if there is no actual call to those methods from the hive side ? that would be misleading thought ?

**Comment by b-slim** [comment]
@jcamachor i have addressed all the comments please check this out.

**Comment by b-slim** [comment]
@jcamachor I AM not against adding it, but what i am saying if we do add it to the interface will need to add the call for it in hive core it self otherwise it will not make sense right ?

**Comment by jcamachor** [comment]
@b-slim , sure, that is what I was pointing out. It should be in the method insertCommit in HiveMetaStoreClient.java.

**Comment by b-slim** [comment]
merged

**Pull request closed without merging**